### PR TITLE
Load and unload tiles proposal

### PR DESCRIPTION
**Pull request by hann2**

### CODE REVIEW

File: src/core/util/SubGrid.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    public y: number,
+    public width: number,
+    public height: number
+  ) {
+
+  }
+
+  has(cell: TilePos): boolean {
+    const [x, y] = cell;
+    return x >= this.x && y >= this.y && x < this.x + this.width && y < this.y + this.width;
```
Comment (hann2) [inline src/core/util/SubGrid.ts]: Bug: Need to change last this.width to this.height

File: src/core/util/SubGridSet.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  [Symbol.iterator](): Iterator<TilePos> {
+    return this.values();
+  }
+
+  * cellGenerator(): Generator<TilePos> {
+    for (let i = 0; i < this.data.length; i++) {
+      const subGrid = this.data[i];
+      for (let x = subGrid.x; x < subGrid.x + subGrid.width; x++) {
+        for (let y = subGrid.y; y < subGrid.y + subGrid.height; y++) {
```
Comment (hann2) [inline src/core/util/SubGridSet.ts:24]: I could make subgrid iterable and remove the x/y for loops.

File: src/mariana/world/WorldMap.ts [hunk 3]
```diff
@@ -117,55 +106,32 @@ export class WorldMap extends BaseEntity implements Entity {
...
-    // get tilesToLoad
-    const tilesToLoad = this.getAnchoredTiles();
-
-    // put tiles loaded into set tilesToUnload
-    const tilesToUnload = new GridSet();
-    for (const tile of this.tilesLoaded) {
-      tilesToUnload.add(tile);
-    }
+    const lastFramesTiles : SubGridSet = this.tilesLoaded;
+    const thisFramesTiles : SubGridSet = this.getAnchoredTiles();
```
Comment (simonbw) [inline src/mariana/world/WorldMap.ts]: I would change "this" to "current".

File: src/mariana/world/WorldMap.ts [hunk 4]
```diff
@@ -117,55 +106,32 @@ export class WorldMap extends BaseEntity implements Entity {
...
 
-    // for each tile of tilesToLoad
-    for (const tile of tilesToLoad) {
-      //   remove tile from tilesToUnload
-      tilesToUnload.delete(tile);
-      //   if tile isn't loaded: load it
-      if (!this.tilesLoaded.has(tile)) {
-        this.loadTile(tile);
+    for (const tile of thisFramesTiles) {
```
Comment (simonbw) [inline src/mariana/world/WorldMap.ts]: This does come out very clean, I like it.

File: src/core/util/SubGridSet.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+              continue;
+            }
+          }
+          yield cell;
+        }
+      }
+    }
+  }
+
+  values(): Iterator<TilePos> {
```
Comment (simonbw) [inline src/core/util/SubGridSet.ts:39]: You can just make `*values()` itself a generator so you don't have to have a separate `*cellGenerator` function.

**Review by simonbw — COMMENTED** [review]
I like it. Performance seems way better, and I think this will turn out to be fast enough for anything we want to do with this game. A couple notes:

- If I dive deep enough in the game, eventually tiles stop loading. I have a feeling it's using meters where it should be using tiles somewhere. I'll try to find the spot.
 - I think all the subgrid stuff is specific enough to this game that it doesn't really deserve a place in `core/` right now, so I would move all that stuff over to somewhere in `mariana/world/`.

File: src/core/util/SubGridSet.ts (hunk 2 above)
Comment (simonbw) [reply src/core/util/SubGridSet.ts:24]: You could even use the `yield*` keyword then!
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield*

I just learned about it today.

File: simons-notes.txt [hunk 6]
```diff
@@ -155,6 +155,17 @@
        - empty ocean
        - coral reef
 
+    Some cool looking coral reef species for inspiration (coral, anemones, & urchins):
```
Comment (simonbw) [inline simons-notes.txt:158]: These are great

**Pull request merged**